### PR TITLE
[SYCL][ESIMD][EMU] Marking ESIMD kernels for esimd_emulator backend

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
This PR is to mark ESIMD kernels for esimd_emulator backend with `UNSUPPORTED` or `XFAIL` for future feature bring-up and debug.

### CODE REVIEW

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  float *A = new float[Size];
+  float *B = new float[Size];
+  float *C = new float[Size];
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = B[i] = i;
+    C[i] = 0.0f;
+  }
+
+  try {
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:36]: wouldn't an empty kernel run via `single_task` (or `parallel_for` if single_task is not supported yet) suffice here?

**Comment by dongkyunahn-intel** [comment]
This PR will be merged after ESIMD_EMULATOR support is enabled by default in intel/llvm - https://github.com/intel/llvm/pull/5058.

File: SYCL/ESIMD/sycl_esimd_mix.cpp [hunk 2]
```diff
@@ -10,9 +10,11 @@
 
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// TODO/FIXME: esimd_emulator support - timeout
+// XFAIL: esimd_emulator
```
Comment (kbobrovs) [inline SYCL/ESIMD/sycl_esimd_mix.cpp]: Changes in this test completely defeat the test's purpose. I think for now, we should just add `UNSUPPORTED: esimd_emulator`, and leave the rest as is.
Eventually, we might need to figure out a way to execute non-ESIMD kernels via IGC and ESIMD kernels - via esimd_emulator.

File: SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
+//==--- sycl_esimd_mix_check_build_opts.cpp - DPC++ ESIMD on-device test ---==//
```
Comment (kbobrovs) [inline SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp]: according to my previous comment, this test is not needed

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: SYCL/lit.cfg.py [hunk 4]
```diff
@@ -189,7 +189,8 @@
                      'opencl',
                      'cuda',
                      'hip',
-                     'level_zero']
+                     'level_zero',
+                     'esimd_emulator']
```
Comment (vladimirlaz) [inline SYCL/lit.cfg.py]: If you have no objections, could we use intel/llvm-test-suite#752 PR to enable esimd_emulator in LIT infra. This PR can be used to have fixes to tests.

Comment (dongkyunahn-intel) [reply SYCL/lit.cfg.py]: I have no objections - more than that I agree. I'll revert this change.

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 1 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp:36]: I'm not sure if it is right to modify this test as it is duplicate of `vadd_1d.cpp`. Original `vadd_1d.cpp` is changed to remove `CHECK` commands for `piProgramBuild` and to mark `XFAIL: esimd_emulator`. This test has those `CHECK` commands and `UNSUPPORTED: esimd_emulator` as `esimd_emulator` does not support online compilation

Do you want this test to be removed like `sycl_esimd_mix_check_build_opts.cpp` and original `vadd_1d.cpp` to have `UNSUPPORTED: esimd_emulator`?

File: SYCL/ESIMD/sycl_esimd_mix.cpp (hunk 2 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/sycl_esimd_mix.cpp]: Will do.

File: SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp (hunk 3 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp]: Will remove.

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  float *A = new float[Size];
+  float *B = new float[Size];
+  float *C = new float[Size];
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = B[i] = i;
+    C[i] = 0.0f;
+  }
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:35]:
```suggestion
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    buffer<float, 1> bufa(A, range<1>(Size));
+    buffer<float, 1> bufb(B, range<1>(Size));
+    buffer<float, 1> bufc(C, range<1>(Size));
+
+    // We need that many workgroups
+    range<1> GlobalRange{Size / VL};
+
+    // We need that many threads in each group
+    range<1> LocalRange{1};
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:46]:
```suggestion
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+          GlobalRange * LocalRange, [=](id<1> i) SYCL_ESIMD_KERNEL {
+            using namespace sycl::ext::intel::experimental::esimd;
+            unsigned int offset = i * VL * sizeof(float);
+            simd<float, VL> va;
+            va.copy_from(PA, offset);
+            simd<float, VL> vb;
+            vb.copy_from(PB, offset);
+            simd<float, VL> vc = va + vb;
+            vc.copy_to(PC, offset);
+          });
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:66]:
```suggestion
  cgh.single_task<class Test>([]{});
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+
+    delete[] A;
+    delete[] B;
+    delete[] C;
+    return 1;
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:75]:
```suggestion
    return 1;
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+              << ((float)(Size - err_cnt) / (float)Size) * 100.0f << "% ("
+              << (Size - err_cnt) << "/" << Size << ")\n";
+  }
+
+  delete[] A;
+  delete[] B;
+  delete[] C;
+
+  std::cout << (err_cnt > 0 ? "FAILED\n" : "Passed\n");
+  return err_cnt > 0 ? 1 : 0;
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:99]:
```suggestion
  return 0;
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 1 above)
Comment (kbobrovs) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp:36]: see suggested changes

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+int main(void) {
+  try {
+    queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+    auto dev = q.get_device();
+    std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+    auto e = q.submit([&](handler &cgh) {
+  cgh.single_task<class Test>([]{});
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp]: please fix formatting

File: SYCL/ESIMD/vadd_1d.cpp [hunk 11]
```diff
@@ -7,8 +7,10 @@
 //===----------------------------------------------------------------------===//
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// TODO: esimd_emulator support - enable __esimd_oword_ld_unaligned
```
Comment (kbobrovs) [inline SYCL/ESIMD/vadd_1d.cpp:10]:
```suggestion
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/esimd_check_vc_codegen.cpp:71]:
```suggestion
```

File: SYCL/ESIMD/vadd_1d.cpp (hunk 11 above)
Comment (kbobrovs) [inline SYCL/ESIMD/vadd_1d.cpp]:
```suggestion
// TODO: esimd_emulator fails due to unimplemented __esimd_oword_ld_unaligned
```

File: SYCL/ESIMD/vadd_1d.cpp [hunk 13]
```diff
@@ -97,6 +99,5 @@ int main(void) {
   return err_cnt > 0 ? 1 : 0;
 }
 
-// CHECK: ---> piProgramBuild(
-// CHECK: <const char *>: {{.*}}-vc-codegen
-// CHECK: ) ---> pi_result : PI_SUCCESS
+// 'CHECK' commands for checking 'vc-codegen' are moved to
+// 'esimd_check_vc_codegen.cpp'
```
Comment (kbobrovs) [inline SYCL/ESIMD/vadd_1d.cpp]:
```suggestion
```

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 10 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp]: Done.

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 12 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp:71]: Done.

File: SYCL/ESIMD/vadd_1d.cpp (hunk 11 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/vadd_1d.cpp]: Done.

File: SYCL/ESIMD/vadd_1d.cpp (hunk 13 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/vadd_1d.cpp]: Done.

File: SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp (hunk 3 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp]: Done.

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 1 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp:36]: Suggestions applied.

File: SYCL/ESIMD/histogram_256_slm_spec_2020.cpp [hunk 14]
```diff
@@ -1,6 +1,7 @@
 // TODO enable on Windows
 // REQUIRES: linux && gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
Comment (kbobrovs) [inline SYCL/ESIMD/histogram_256_slm_spec_2020.cpp:5]: TODO?

File: SYCL/ESIMD/mandelbrot/mandelbrot_spec.cpp [hunk 15]
```diff
@@ -9,6 +9,7 @@
 // TODO enable on Windows
 // REQUIRES: linux && gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
Comment (kbobrovs) [inline SYCL/ESIMD/mandelbrot/mandelbrot_spec.cpp:13]: TODO?

File: SYCL/ESIMD/regression/complex-lib-lin.cpp [hunk 16]
```diff
@@ -7,6 +7,7 @@
 //
 // REQUIRES: linux,gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
Comment (kbobrovs) [inline SYCL/ESIMD/regression/complex-lib-lin.cpp:11]: TODO?

File: SYCL/ESIMD/regression/globals.cpp [hunk 17]
```diff
@@ -1,5 +1,7 @@
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// Sub-group is not supported
```
Comment (kbobrovs) [inline SYCL/ESIMD/regression/globals.cpp]:
```suggestion
// TODO Sub-group is not supported
```

File: SYCL/ESIMD/spec_const/spec_const_bool.cpp [hunk 18]
```diff
@@ -9,6 +9,7 @@
 // RUN: %clangxx -fsycl -I%S/.. %s -o %t.out
 // RUN: %GPU_RUN_PLACEHOLDER %t.out
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
Comment (kbobrovs) [inline SYCL/ESIMD/spec_const/spec_const_bool.cpp:13]: TODO?  here and other places. We should eventually support spec constants in the emulator.

**Review by kbobrovs — DISMISSED** [review]

File: SYCL/ESIMD/spec_const/spec_const_bool.cpp (hunk 18 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/spec_const/spec_const_bool.cpp:13]: Like I showed in messenger, online_compiler check fails for all spec* tests - even if it is not required. To whom should I ask about this issue? (Someone in SYCL_RT?)

File: SYCL/ESIMD/esimd_check_vc_codegen.cpp (hunk 9 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_check_vc_codegen.cpp:99]: @kbobrovs , `esimd_check_vc_codegen.cpp` fails after being simplified with `single_task` - for both open_cl and level_zero. Maybe it has to be reverted.

```
[2022-01-27T05:52:29.296Z] /netbatch/donb02499_00/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm-test-suite/SYCL/ESIMD/esimd_check_vc_codegen.cpp:41:11: error: CHECK: expected string not found in input
[2022-01-27T05:52:29.296Z] // CHECK: <const char *>: {{.*}}-vc-codegen
[2022-01-27T05:52:29.296Z]           ^
[2022-01-27T05:52:29.296Z] <stdin>:415:2: note: scanning from here
[2022-01-27T05:52:29.296Z]  <unknown> : 0x7056b0
[2022-01-27T05:52:29.296Z]  ^
[2022-01-27T05:52:29.296Z] <stdin>:418:2: note: possible intended match here
[2022-01-27T05:52:29.296Z]  <const char *>: 
[2022-01-27T05:52:29.296Z]  ^
[2022-01-27T05:52:29.296Z] 
```

**Comment by dongkyunahn-intel** [comment]
`esimd_check_vc_codegen.cpp` will be revised to minimized test as @kbobrovs suggested in a separate PR.

**Review by kbobrovs — DISMISSED** [review]
LGTM, but please address the vc_code_gen test comment at your earliest convenience.

**Review by vladimirlaz — APPROVED** [review]

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**